Memory backed api

joe-iddon opened the pull request:
<!--
Thanks for contributing a Pull Request to ArcticDB! Please ensure you have taken a look at:
 - ArcticDB's Code of Conduct: https://github.com/man-group/ArcticDB/blob/master/CODE_OF_CONDUCT.md
 - ArcticDB's Contribution Licensing: https://github.com/man-group/ArcticDB/blob/master/docs/mkdocs/docs/technical/contributing.md#contribution-licensing
-->

#### Reference Issues/PRs

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged.

See: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

#### What does this implement/fix? How does it work (high level)? Highlight notable design decisions.

Implements a memory-backed API for the top-level `Arctic()` instance.

Intended usage:

```python
>>> import pandas as pd
>>> ac = Arctic("mem://memory_db") # this name is passed through to the C++ storage, but ignored there
>>> ac.create_library("lib")
...
```

Test definition files which now also verify `mem://`:
- `test_arctic_batch.py` - added via extending `arctic_library`/`arctic_client`
- `test_arctic.py` - added again via `arctic_library`/`arctic_client`
- `test_basic_version_store.py` - parameterised all `lmdb_version_store` fixtures to test the memory adapter too*
- `test_stress_generated.py` - parameterised `lmdb_version_store_tiny_segment` to test the memory adapter too.

\* I did not extend the more specific test fixtures like `lmdb_version_store_tombstone_and_sync_passive`.

These add an extra 15 seconds on top of times around 3 mins for the large test files.

#### Any other comments?

The timings for the stress test (`test_stress_small_row` in `test_stress_generated.py`) are:


Adapter type | Write / sec | Read / sec
--- | --- | ---
lmdb | 97 | 5
mem | 11 | 10

The test is for 1MB of data. The results for the high entropy and low entropy times were within a few seconds of each other. (High entropy has completely random data, low entropy has blocks of repeated rows: aaacccbbb.)

At the moment the name given is unused.

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x ] Have you updated the relevant docstrings and documentation?
 - [x ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [x ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [x ] Are API changes highlighted in the PR description?
 - [x ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

poodlewars commented:
Please rebase and squash the commits just before merging this one

poodlewars commented on python/arcticdb/arctic.py:
```diff
@@ -140,6 +147,17 @@ def __init__(self, uri: str, encoding_version: EncodingVersion = DEFAULT_ENCODIN
+
+            In-Memory
+            ----
+                The in-memory connection scheme has the form ``mem://<name>`, e.g. `mem://memory_db`.
```
I wonder what the point of `<name>` is then? For now, all I can it doing is adding confusion. Shall we just have `mem://`?

jamesmunro replied:
- [ ] remove name

poodlewars commented on python/tests/integration/arcticdb/test_in_memory.py:
What's the point of this one? Doesn't it cover stuff already tested in `test_arctic.py` with the new fixture parameter you added?

joe-iddon replied:
We discussed that the `arctic_client` in `conftest.py` actually does verify that this works already, but it is good to have an explicit high-level check anyway.

poodlewars commented on python/tests/integration/arcticdb/version_store/test_basic_version_store.py:
This conflicts with @G-D-Petrov 's PR #789 . https://github.com/man-group/ArcticDB/pull/789/files#diff-820e1d6f588509b49c8841acba4cc4c430a434a90a8fa483130ed6af36076048

Probably just calling your new fixture `basic_store` instead will help make the rebase easier for one of you.

Please talk to Georgi to figure out how / when you two can merge your changes.

poodlewars commented:
After this is merged we should add some documentation about it to the website docs.arcticdb.io

jamesmunro commented:
Why slower to write?  This suggests something is wrong.

- [ ] explanation for performance

<html>
<body>
<!--StartFragment-->

Adapter type | Write / sec | Read / sec
-- | -- | --
lmdb | 97 | 5
mem | 11 | 10

<!--EndFragment-->
</body>
</html>

jamesmunro commented on python/arcticdb/adapters/in_memory_library_adapter.py:
```diff
@@ -0,0 +1,92 @@
+    Supports any URI that begins with `mem://` - for example, `mem://in_memory_db`.
+    """
+
+    # Only query parameter is cache_size at the moment, but this is not even used.
```
- [ ] remove the code to handle the query parameters

joe-iddon commented on cpp/arcticdb/storage/memory/memory_storage-inl.hpp:
```diff
@@ -31,13 +30,21 @@ namespace arcticdb::storage::memory {
                 util::variant_match(kv.variant_key(),
                                     [&](const RefKey &key) {
-                                        key_vec[key] = kv.segment();
+                                        auto it = key_vec.find(key);
```
Keeps the `it` variable just in the scope of the if-statement. 👍

joe-iddon commented:
> Why slower to write? This suggests something is wrong.

Lmdb is presumably slower because it is creating and writing to files whereas the in-memory is just a `std::unordered_map` (really a folly `ConcurrentHashMap`). I'm going to document the in-memory API in another PR after merging this one, and will be investigating lmdb on tmpfs which I can compare timings for against lmdb on a real directory.

I don't think my code being faster suggests anything is wrong.

joe-iddon commented:
Re: performance tests. Sorry @jamesmunro - the table headers are confusing. They are not writes-per-second, but the time taken, in seconds, to complete the writing test. Good to have some explanations anyway.

joe-iddon commented:
Some more timings data (again averaging high entropy and low entropy discrepancies can be ignored):

With 1MB of data:

Adapter type | Write / sec | Read / sec
-- | -- | --
lmdb | 25 | 4
mem | 10 | 4

With 2MB of data:

poodlewars commented on cpp/proto/arcticc/pb2/in_memory_storage.proto:
```diff
@@ -10,7 +10,6 @@ syntax = "proto3";
 package arcticc.pb2.in_memory_storage_pb2;
 
 message Config {
-    uint64 cache_size = 1; // if 0 => 10GB
```
Scary to change a protobuf in a backwards incompatible way!

In this case probably fine, I doubt anyone has serialized this type of config anywhere. Doesn't seem worth the risk though, suggest just adding a comment to explain that this field is deprecated and unused instead?

poodlewars commented on python/arcticdb/adapters/in_memory_library_adapter.py:
```diff
@@ -0,0 +1,71 @@
+    """
+    Use in-memory storage.
+
+    Supports only the URI `"mem://"`.
```
```suggestion
    Only supports the URI `"mem://"`.
```

is a bit more grammatical

joe-iddon replied:
Agreed. I'll make the change in the docs PR.

poodlewars commented on python/arcticdb/arctic.py:
```diff
@@ -138,6 +145,14 @@ def __init__(self, uri: str, encoding_version: EncodingVersion = DEFAULT_ENCODIN
+            ----
+                The in-memory connection scheme has the form ``"mem://"`.
+
+                The storage is local to the returned `Arctic` instance.
```
Double backticks around `Arctic`?

joe-iddon replied:
Thanks. Changes are in the docs PR here: https://github.com/man-group/ArcticDB/pull/895/files#diff-a9727c85029e3b2014647019b25f48df66c2670da1d17f0e2f7e279242f0a8a6R147-R149

poodlewars commented on python/arcticdb/arctic.py:
```diff
@@ -138,6 +145,14 @@ def __init__(self, uri: str, encoding_version: EncodingVersion = DEFAULT_ENCODIN
+
+            In-Memory
+            ----
+                The in-memory connection scheme has the form ``"mem://"`.
```
mismatched backticks

joe-iddon replied:
Changes are in the docs PR here: https://github.com/man-group/ArcticDB/pull/895/files#diff-a9727c85029e3b2014647019b25f48df66c2670da1d17f0e2f7e279242f0a8a6R147-R149

joe-iddon commented:
For the first merge from master which bought @G-D-Petrov's changes, I had to make these changes to resolve the many merge conflicts: (These are also in the commit message)

  * test_restore_version causes a duplicate SYMBOL_LIST AtomKey. LMDB was silently dealing with this
    by returning a DuplicateKeyException, but in-memory storage was catching this and throwing an error.
    --> Modified in-memory storage to also throw a DuplicateKeyExcpetion like LMDB, so that the
        write_journal method in symbol_list.cpp will catch this and print a debug message silently,
        as is the behaviour of LMDB.
  * fixed lmdb_config being defined in conftest.py for all test types, including things which don't
    know how to ignore it (used Georgi's hack to remove from kwargs before setting the cfg)
  * added in-memory store to the unit tests which tested all backends (in line with Georgi's changes
    as we had agreed in advance for this merge)
  * fixed flaky test_separation_between_libraries test which was comparing lists, not sets
  * added a dynamic_schema unit test method for in-memory store to match the others

joe-iddon commented:
(Brought in Alex's changes from master to hopefully fix macos pytest fails.)

poodlewars commented:
Please rebase before merging and

- Squash commits in to one
- Do not include the merge commits
